snapshots: Fix snapshot-ctl to include snapshots of dropped tables

bhalevy opened the pull request:
Snapshot-ctl methods fetch information about snapshots from
column family objects. The problem with this is that we get rid
of these objects once the table gets dropped, while the snapshots
might still be present (the auto_snapshot option is specifically
made to create this kind of situation). This commit switches from
relying on column family interface to scanning every datadir
that the database knows of in search for "snapshots" folders.

This PR is a rebased version of #9539 (and slightly cleaned-up, cosmetically)
and so it replaces the previous PR.

Fixes #3463
Closes #7122

nyh commented:
@bhalevy this is a single-patch pull-request, so its cover letter will be lost - so please move what only exists in the cover letter (namely, the Fixes, Closes) in the patch's commit message.
Also, please rebase - there are conflicts.

bhalevy commented:
> @bhalevy this is a single-patch pull-request, so its cover letter will be lost - so please move what only exists in the cover letter (namely, the Fixes, Closes) in the patch's commit message. Also, please rebase - there are conflicts.

@nyh: done

nyh commented:
@bhalevy github got confused on where I put my comments, but please take a look at them even though they are marked "outdated".

kbr- commented on db/snapshot-ctl.cc:
```diff
@@ -144,82 +146,58 @@ future<> snapshot_ctl::clear_snapshot(sstring tag, std::vector<sstring> keyspace
-                });
-            });
+    co_return co_await run_snapshot_list_operation([this] () -> future<int64_t> {
+        co_return co_await _db.map_reduce(adder<int64_t>(), [] (replica::database& db) -> future<int64_t> {
```
One thing I don't understand (and wanted to test before merging the previous PR, but didn't get to it) is why we have to reduce over all shards? It seems that `db.get_snapshot_details()` should already return all snapshots already on one shard (any shard), since it scans all the directories.

That said, I also don't understand why the previous version had to reduce over shards.

bhalevy replied:
I agree, the snapshot directory is shared.
Will fix.

kbr- replied:
Does it mean that the previous version was buggy as well?

bhalevy replied:
Yes, `snapshot_ctl::true_snapshots_size` is inflated in a factor of `smp::count`.
We may want to fix that independently and backport the fix (with no coroutines).
Let's open an issue and I'll extract the patch fixing it so it can be backported.

bhalevy replied:
https://github.com/scylladb/scylla/issues/9897

bhalevy commented on database.cc:
```diff
@@ -2150,6 +2150,70 @@ static sstring get_snapshot_table_dir_prefix(const sstring& table_name) {
 }
 
+static sstring extract_cf_name(const sstring& directory_name) {
+    // cf directory is of the form: 'cf_name-timestamp'
```
could be a timeuuid :)
Anyhow, I'll fix the comment.

bhalevy commented on database.cc:
```diff
@@ -2150,6 +2150,70 @@ static sstring get_snapshot_table_dir_prefix(const sstring& table_name) {
 
+static sstring extract_cf_name(const sstring& directory_name) {
+    // cf directory is of the form: 'cf_name-timestamp'
+    return directory_name.substr(0, directory_name.find("-"));
```
Agreed regarding finding last of `'-'`
`distributed_loader::populate_keyspace` gets the table names from the schema and the uuid suffix from `schema::id()` so it doesn't parse the table name but rather it reconstructs it form the available metadata.

sstables `make_entry_descriptor` doesn't care about the table name so the regular expression it uses doesn't separate the table name from the uuid

bhalevy commented on database.cc:
```diff
@@ -2150,6 +2150,70 @@ static sstring get_snapshot_table_dir_prefix(const sstring& table_name) {
+            sstring ks_name = de.name;
+
+            co_return co_await lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) -> future<> {
+                // CF directory
```
It's orthogonal to the sstables version.
There shouldn't be any dependency even on scylla version at least form version 3.0 IIRC.

nyh replied:
I seem to have a vague recollection that the directory structure **was** different in old sstable versions - subdirectories were introduced for individual tables where previously all the tables were in the same per-keyspace directory. I'm not sure I'm remembering this right...

bhalevy replied:
Hmm, this is probably from the `ka` times apparently,
since they hold the keyspace and column_family name in the sstable name so they can all be placed in a shared data directory.

See https://github.com/scylladb/scylla/blob/master/docs/design-notes/sstables-directory-structure.md
```
    <keyspace>-<column_family>-ka-<generation>-<component>
```

nyh replied:
Right. So my question is - what does this patch do in this case? Does it at least do something reasonable (even if not perfect)?

bhalevy replied:
> Right (so I mixed up the details in my statement above). So my question is - what does this patch do in this case? Does it at least do something reasonable (even if not perfect)?

This patch or the present state before it AFAICT don't handle the shared-data-directory-with-no-pre-table-sub-directories case as they expect the snapshots subdirectory to live under the per-table directory.
Frankly, I don't know if we even supported snapshots in the shared data directory times...

bhalevy commented:
In v2:
- Simplified `snapshot_ctl::get_snapshot_details` and `snapshot_ctl::true_snapshots_size`
  - no need for map_reduce across all shards
- database `extract_cf_name`
  - fixed comment regarding uuid
  - `find_last_of('-')` as the table name might contain a `'-'`
    - throw internal error if no `'-'` is found.
- `database::get_snapshot_details`
  - simplified and optimized a bit - explicitly check for a `sstables::snapshots_dir` rather than using a filter functor
    - other than tables with snapshots we also visited non-keyspace subdirectoris under the data dir like commitlog or view_hints which contain subdirectories of their own that aren't even table directories.
- when finding a file with same name both in the table directory and the snapshots sub-directory make sure they both hardlinks to he same inode, otherwise log a warning and add size to "live" counter.

bhalevy commented:
Let's get https://github.com/scylladb/scylla/pull/9899 merged first as we may want to backport it
and I'll rebase this PR on top of it.

kbr- commented:
But this PR was already queued.

bhalevy commented:
> But this PR was already queued.

Ooops.  @nyh what do you think?
This PR fixes #9897 too, and we can use d346f0ad57e711fafda0d1f8423f0e9214417bf4 for backporting to 4.6 and below.
e721501067f8f51dd3b7111d274aec8938de823d can also be applied on top of this PR and it is less important for backporting.

nyh commented:
> > But this PR was already queued.
> 
> Ooops. @nyh what do you think?

I don't know, tell me what you want to do (dequeue it, or not). I don't understand why the order of the various fixes is important or effects the backporting, but you have a better view on this. You never mentioned before that you *didn't* want to commit this patch because you're waiting for something else first.

bhalevy commented:
> I don't know, tell me what you want to do (dequeue it, or not). I don't understand why the order of the various fixes is important or effects the backporting, but you have a better view on this. You never mentioned before that you _didn't_ want to commit this patch because you're waiting for something else first.

Sorry, I think it'd be better to dequeue it in favor of merging #9899 first.
It's easier for backporting since the fixes in the other PR are based on code much similar if not identical to the code in the release branches.

bhalevy commented:
> You never mentioned before that you didn't want to commit this patch because you're waiting for something else first.

I realized that pretty much at the same time you merged this PR (and thanks!)

nyh commented:
Dequeued for now. Please change this PR to "draft" status, and only un-draft it when you want it committed.

bhalevy commented:
In v3:
- rebased + retested
  - unit(dev) + database_test.snapshot_list_contains_dropped_tables(debug)

bhalevy commented:
@nyh please re-review and merge if there are no objections

bhalevy commented:
ping @nyh
